# ハンバーガー統計学 with Python 第6章

第５回講義（2017年10月13日講義、10月17日課題提出締切）

このページは、<a href="http://kogolab.chillout.jp/elearn/hamburger/index.html" target="_blank">ハンバーガー統計学にようこそ！</a>の内容をプログラミング言語 Python に合わせて書き直したものです（著作権者の許諾を得て利用しております）。

<h1 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap6/sec0.html" target="_blank">第6章 分散分析（１要因）</a></h1>

<img src="image/salesman00021.jpg">

「困ったよ～！！！！」

あなたがワクワクバーガーのお店に行くと、あなたの姿を見つけたとたん、店員さんが走ってきました。

新しく作られたハンバーガーは、実際に新商品として売り出されて、順調に売り上げを伸ばしていると聞いていたあなたは、一体何が起きたのか気になりました。

「実は、このお店とモグモグの間に、新しいハンバーガーショップ、バーガーパクパク亭が開店したんだ。ただでさえ、二店でしのぎをけずってたのに、3店じゃもう……」

<h2 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap6/sec1.html" target="_blank">6.1　なぜt検定が使えないか</a></h2>

バーガーパクパク亭という、新しいライバル店の出現です。

<table border="0"><tr>
<td><img src="image/m-shop.jpg"></td>
<td><img src="image/p-shop.jpg"></td>
<td><img src="image/w-shop.jpg"></td>
</tr></table>

あなたは店員さんからこんなお願いをされました。

「とりあえずうちは、ポテトが看板商品だから、t検定をやって、3つのどのポテトが一番おいしいか調べてくれよ～～～～。」

<table border="0"><tr>
<td><img src="image/m-potato00001.jpg"></td>
<td><img src="image/p-potato.jpg"></td>
<td><img src="image/w-potato00001.jpg"></td>
</tr></table>

「それはできません」

「えっ……（絶句）」

<img src="image/salesman00004.jpg">

あなたにできないことはないと思っていた店員さんは絶句しました。

「あ、いや、自分ができない、ってことじゃなくて、t検定では、それはできない、ってことです。t検定は、2つの標本間の平均の差を調べるものなので、標本が3つ以上になると、t検定は使っちゃダメなんです。今回の場合だと、ワクワク、モグモグ、パクパクの3つですよね」

標本が3つ以上だとt検定は使えない？

「え、だって、ワクワクとモグモグ、モグモグとパクパク、ワクワクとパクパク、この3つの組み合わせについてt検定をやればいい話じゃないの？」

「あの、それがダメなんです。やっちゃいけないんです」

<h4 style="border-bottom: solid 1px black;">なぜt検定が使えないか</h4>

t検定の話に入る前に、コイン投げの話をします。

「コインを1回投げて、表が出る確率は？」

これは簡単ですね。1/2です。

 

「では、コインを2回投げて、少なくとも1回表が出る確率は？」

「少なくとも1回表が出る」というのは、「全部裏が出る」の反対の事象になります。

少なくとも1回表が出る　＝　（表　と　表）、（表　と　裏）、（裏　と　表）
全部裏が出る　　　　　　＝　（裏　と　裏）
つまり「少なくとも1回表が出る」確率は、全部の事象（確率は1）から2回とも裏が出る確率をひいたものになります。

<img src="image/eq6001coin.gif">

つまり、2回コインを投げると、少なくとも1回表が出る確率は、0.75となり、1回しか投げなかったときの確率0.5よりも高くなりますね。

 

「では、サイコロを2回振って、少なくとも1回は3の目がでる確率は？」

これは、全部の事象（確率1）から2回とも3以外の目が出る確率をひいたものになります。

<img src="image/eq6002dice.gif">

つまり、2回サイコロを振ると、少なくとも1回3の目が出る確率は、0.31となり、1回しか振らなかったときの確率（1/6＝0.17）よりも高くなりますね。

 

このように、同じことを繰り返すと、特定のことが起こる確率が高くなるということがわかります。

そこで、t検定の話に入ります。

A, B, Cの3つの標本があるとき、AとB、BとC、AとCの、3つにおいてt検定を行った場合、「少なくとも1つの組み合わせに、差がでる確率」は、1から「3つの組み合わせすべてに差がでない確率」をひいたものになることが分かります。

「差がでない確率」は、全体の確率（1）から差がでる確率をひいたものになります。ここでは、差が出る確率を5％（0.05）としましょう。そうすると、差がでない確率は、1－0.05　となります。

そうすると「少なくとも1つの組み合わせに、差がでる確率」は次のようになります。

<img src="image/eq6003sa.gif">

これを計算すると、

＝　1　－　（0.95×0.95×0.95）
＝　1　－　0.857　
＝　0.143

となります。

「少なくとも1つの組み合わせに、差がでる確率」は、0.143になり、1回だけのときの確率（0.05）に比べて、差がでる確率が高くなります。つまり、0.05から比べると3倍弱、確率が高くなります。

つまり、比較する回数が増えれば増えるほど、実際は差がないのに、差があるとされる確率が増えてしまうことになります。

これが、t検定は3つ以上の標本間の差の検定には使えないという理由です。

<h2 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap6/sec2.html" target="_blank">6.2　分散分析とは</a></h2>

ここではt検定は使えないので、別の検定方法を考えなければなりません。その方法は「分散分析」といいます。

分散分析について説明する前に、まずデータを取ってきましょう。

ワクワク、モグモグ、パクパクのポテト20個ずつを手に入れて、駅に向かい、通行人合計60人にランダムにどれかひとつのポテトを食べてもらいました。そして、そのおいしさについて、100点満点で点数をつけてもらいました。

以下が、それぞれのお店のポテトの評価データです。

In [1]:
wakuwaku = [80, 75, 80, 90, 95, 80, 80, 85, 85, 80, 90, 80, 75, 90, 85, 85, 90, 90, 85, 80]
mogumogu = [75, 70, 80, 85, 90, 75, 85, 80, 80, 75, 80, 75, 70, 85, 80, 75, 80, 80, 90, 80]
pakupaku = [80, 80, 80, 90, 95, 85, 95, 90, 85, 90, 95, 85, 98, 95, 85, 85, 90, 90, 85, 85]

それぞれのお店のデータ数と平均と標準偏差を計算しましょう。

In [2]:
# 平均値を求める関数
def average(list):
    sum = 0
    for x in list:
        sum += x
    return sum / len(list)

In [3]:
# 分散を計算する
def variance(list):
    ave = average(list)
    sum = 0
    for x in list:
        sum += (x - ave) ** 2
    return sum / len(list)

In [4]:
# 標準偏差を計算する
import math
def standard_deviation(list):
    return math.sqrt(variance(list))

In [5]:
# それぞれのお店のデータ数と平均と標準偏差
print(len(wakuwaku), average(wakuwaku), standard_deviation(wakuwaku))
print(len(mogumogu), average(mogumogu), standard_deviation(mogumogu))
print(len(pakupaku), average(pakupaku), standard_deviation(pakupaku))

20 84.0 5.385164807134504
20 79.5 5.454356057317857
20 88.15 5.341114115987412


さらに、全体の、データ数と平均と標準偏差を計算しましょう。

In [6]:
# 全体のデータ数と平均と標準偏差
zentai = wakuwaku + mogumogu + pakupaku
print(len(zentai), average(zentai), standard_deviation(zentai))

60 83.88333333333334 6.447458586313078


計算結果を見ると、標準偏差はどの店もだいたい同じ。評価の平均は、パクパク（88.15）、ワクワク（84.00）、モグモグ（79.50）の順で高いようです。また、全体の平均は、83.88でした。

<h4 style="border-bottom: solid 1px black;">帰無仮説、対立仮説を立てよう</h4>

それでは各平均が出揃ったところで、3つ以上の平均の差の検定方法について考えてみましょう。

最初に、帰無仮説を立てることから出発します。

帰無仮説は「3つのお店のポテトの評価（母集団）の平均に差はない」です。

もっと正確に言うと「3つのお店のポテトの評価（母集団）の平均のどの組み合わせにおいても差はない」ということです。

そうすると、対立仮説は「3つのお店のポテトの評価（母集団）の平均の少なくともひとつの組み合わせに差がある」となります。

ここで、対立仮説は「すべての組み合わせに差がある」ではないことに注意してください。そうではなく「少なくともひとつの組み合わせに差がある」ということです。「すべての組み合わせに差がある」場合もこの中に含まれることになります。

<h4 style="border-bottom: solid 1px black;">平均からのズレ（分散）を分析するから分散分析</h4>

それでは、分散分析の考え方について説明していきましょう。

いま、3つのお店の評価データはこんなふうになっています（分布の形はいいかげんです）。

<img src="image/anova2.jpg">

モグモグの中のひとつのデータ（●で示したもの）について考えます。

<img src="image/anova3.jpg">

このデータは、全体の平均から矢印の分だけズレています。

よく見ると、全体の平均からのズレは、「全体の平均とモグモグの平均のズレ」と「モグモグの平均からのズレ」に分解できます。下のように。

<img src="image/anova4.jpg">

ここで、「全体の平均とモグモグの平均のズレ」は何を表しているかと考えると、これは全体の平均から各群（各標本集団）がどれほどズレているかということです。これを「群間のズレ」と呼びましょう。

一方、「モグモグの平均からのズレ」は何を表しているかというと、群（標本集団）の中で、個々のデータがどれほどズレているかということです。これを「群内のズレ」と呼びましょう。

そうすると、すべてのデータについて、全体の平均からのズレは、群間のズレと群内のズレに分解できます。

つまり、すべてのデータについて、

全体の平均からのズレ　＝　群間のズレ　＋　群内のズレ

ということが成り立ちます。

<h4 style="border-bottom: solid 1px black;">群間のズレと群内のズレを比較する</h4>

さて、群間のズレは、標本集団の間の違いを表しています。これが大きくなるということは、各群の平均が大きく異なるということです。

一方、群内のズレは、同じ標本集団の中でのばらつきですので、「誤差」や「個人差」として扱うことができます。

もし、群内のズレに比べて、群間のズレが大きければ、標本集団の間の違いが大きいということですから、「母集団の平均に差がない」という帰無仮説を棄却することになります。

逆に、群内のズレに比べて、群間のズレが小さければ、標本集団の間の違いが大きいとはいえないですから、「母集団の平均に差がない」という帰無仮説を採択することになります。

これが分散分析の考え方です。

<h2 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap6/sec3.html" target="_blank">6.3　分散分析をしよう</a></h2>

では、実際にデータを使って、分散分析の計算をしてみましょう。

<h4 style="padding: 0.25em 0.5em;color: #494949;background: transparent;border-left: solid 5px #7db4e6;">課題6.2</h4>

ワクワク、モグモグ、パクパクのポテト20個ずつを手に入れて、駅に向かい、通行人合計60人にランダムにどれかひとつのポテトを食べてもらいました。そして、そのおいしさについて、100点満点で点数をつけてもらいました。このデータについて、以下の問いに答えてください。
 <TABLE BORDER="1" align="center" CELLPADDING="0" CELLSPACING="0">
    <TR align="center"> 
      <td bgcolor="#cccccc"></td>
      <TD bgcolor="#EEBBBB">ワクワク</TD>
      <TD bgcolor="#BBBBEE">モグモグ</TD>
      <TD bgcolor="#BBEEBB">パクパク</TD>
      <td bgcolor="#cccccc">全体</td>
      <td bgcolor="#cccccc">群内</td>
      <td bgcolor="#cccccc">群間</td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">1</td>
      <TD HEIGHT="30">80</TD>
      <TD>75</TD>
      <TD>80</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">2</td>
      <TD HEIGHT="30">75</TD>
      <TD>70</TD>
      <TD>80</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">3</td>
      <TD HEIGHT="30">80</TD>
      <TD>80</TD>
      <TD>80</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">4</td>
      <TD HEIGHT="30">90</TD>
      <TD>85</TD>
      <TD>90</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">5</td>
      <TD HEIGHT="30">95</TD>
      <TD>90</TD>
      <TD>95</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">6</td>
      <TD HEIGHT="30">80</TD>
      <TD>75</TD>
      <TD>85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">7</td>
      <TD HEIGHT="30">80</TD>
      <TD>85</TD>
      <TD>95</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">8</td>
      <TD HEIGHT="30">85</TD>
      <TD>80</TD>
      <TD>90</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">9</td>
      <TD HEIGHT="30">85</TD>
      <TD>80</TD>
      <TD>85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">10</td>
      <TD HEIGHT="30">80</TD>
      <TD>75</TD>
      <TD>90</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">11</td>
      <TD HEIGHT="30">90</TD>
      <TD>80</TD>
      <TD>95</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">12</td>
      <TD HEIGHT="30">80</TD>
      <TD>75</TD>
      <TD>85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">13</td>
      <TD HEIGHT="30">75</TD>
      <TD>70</TD>
      <TD>98</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">14</td>
      <TD HEIGHT="30">90</TD>
      <TD>85</TD>
      <TD>95</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">15</td>
      <TD HEIGHT="30">85</TD>
      <TD>80</TD>
      <TD>85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">16</td>
      <TD HEIGHT="30">85</TD>
      <TD>75</TD>
      <TD>85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">17</td>
      <TD HEIGHT="30">90</TD>
      <TD>80</TD>
      <TD>90</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">18</td>
      <TD HEIGHT="30">90</TD>
      <TD>80</TD>
      <TD>90</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">19</td>
      <TD HEIGHT="30">85</TD>
      <TD>90</TD>
      <TD>85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR align=center> 
      <td bgcolor="#cccccc">20</td>
      <TD HEIGHT="30">80</TD>
      <TD>80</TD>
      <TD>85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <tr><td bgcolor="#cccccc">平均</td><td>( a )</td><td>( b )</td><td>( c )</td><td>( d )</td>
          <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </tr>
    <tr><td bgcolor="#cccccc">偏差平方和</td><td>( e )</td><td>( f )</td><td>( g )</td><td>( h )</td><td>( i )</td><td>( j )</td></tr>
  </TABLE>

* 帰無仮説を「3つのお店のポテトの評価（母集団）の平均のどの組み合わせにおいても差はない」とします。対立仮説は何か答えてください。

In [7]:
# 対立仮説
# 3つのお店のポテトの評価（母集団）の平均のうち、いずれかの組み合わせにおいて差がある。

* それぞれの店のポテトの点数の平均（上表 a, b, c ）と全体の平均（上表 d ）を求めてください。

In [8]:
wakuwaku = [80, 75, 80, 90, 95, 80, 80, 85, 85, 80, 90, 80, 75, 90, 85, 85, 90, 90, 85, 80]
mogumogu = [75, 70, 80, 85, 90, 75, 85, 80, 80, 75, 80, 75, 70, 85, 80, 75, 80, 80, 90, 80]
pakupaku = [80, 80, 80, 90, 95, 85, 95, 90, 85, 90, 95, 85, 98, 95, 85, 85, 90, 90, 85, 85]

In [9]:
# それぞれの店のポテトの点数の平均（上表 a, b, c ）
a = average(wakuwaku)
b = average(mogumogu)
c = average(pakupaku)
print (a, b, c)

84.0 79.5 88.15


In [10]:
# 全体の平均（上表 d ）
d = average(zentai)
print(d)

83.88333333333334


* それぞれの店のポテトの点数の偏差平方和（上表 e, f, g ）と全体の偏差平方和（上表 h ）を求めてください。

In [11]:
def sum_square_deviation(list): # 偏差平方和
    ave = average(list)
    sum = 0
    for x in list:
        sum += (x - ave)**2
    return sum

In [12]:
# それぞれの店のポテトの点数の偏差平方和（上表 e, f, g ）
e = sum_square_deviation(wakuwaku)
f = sum_square_deviation(mogumogu)
g = sum_square_deviation(pakupaku)
print (e, f, g)

580.0 595.0 570.5499999999998


In [13]:
# 全体の偏差平方和（上表 h ）
h = sum_square_deviation(zentai)
print(h)

2494.1833333333343


* 群内の偏差平方和（上表　i ）を求めてください。群内の平方和＝Σ（それぞれの群の偏差平方和）

In [14]:
# 群内の偏差平方和（上表　i ）
i = sum_square_deviation(wakuwaku) + sum_square_deviation(mogumogu) + sum_square_deviation(pakupaku)
print (i)

1745.5499999999997


* 群間の偏差平方和（上表 j ）を求めてください。 群間の平方和 = Σ((群内平均ー全体平均)の2乗✕個数)

In [15]:
# 群間の偏差平方和（上表 j ）
j = 0
for group in [wakuwaku, mogumogu, pakupaku]:
    j += (average(group) - average(zentai))**2 * len(group)
print (j)

748.6333333333343


<h4 style="border-bottom: solid 1px black;">分散分析表を作る</h4>

分散分析をするために、分散分析表を作ります。群間・群内・全体の偏差平方和は既に計算済みです。
<TABLE BORDER="1" align="center" CELLPADDING="2" CELLSPACING="0">
  <TR> 
    <TD WIDTH=100 BGCOLOR="#CCCCCC"> <CENTER>
        要因 </CENTER></TD>
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
        偏差平方和 </CENTER></TD>
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
        自由度 </CENTER></TD>
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
        平均平方 </CENTER></TD>
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
        F
</CENTER></TD>
  </TR>
  <TR> 
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
        群間
</CENTER></TD>
    <TD WIDTH=100> <CENTER>( j )
        　 </CENTER></TD>
    <TD WIDTH=100> <CENTER>( k )
        　 </CENTER></TD>
    <TD WIDTH=100> <CENTER>( n )
        　 </CENTER></TD>
    <TD WIDTH=100> <CENTER>( p )
        　 </CENTER></TD>
  </TR>
  <TR> 
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
        群内 
      </CENTER></TD>
    <TD WIDTH=100> <CENTER>( i )
        　 </CENTER></TD>
    <TD WIDTH=100> <CENTER>( l )
        　 </CENTER></TD>
    <TD WIDTH=100> <CENTER>( o )
        　 </CENTER></TD>
  </TR>
  <TR> 
    <TD WIDTH=100 BGCOLOR="#CCCCCC"> <CENTER>
        全体 </CENTER></TD>
    <TD WIDTH=100> <CENTER>( h )
        　 </CENTER></TD>
    <TD WIDTH=100> <CENTER>( m )
        　 </CENTER></TD>
  </TR>
</TABLE>

* 自由度を求めてください。
  * ( k ) 群間の自由度　=　群の数　－　1
  * ( l ) 群内の自由度　=　（群1のデータ数－1）＋（群2のデータ数－1）+（群3のデータ数－1）
  * ( m ) 全体の自由度　=　データの総個数　－　1

In [16]:
# 自由度
k = 3 - 1
l = (len(wakuwaku) - 1) + (len(mogumogu) - 1) + (len(pakupaku) - 1)
m = len(wakuwaku + mogumogu + pakupaku) - 1
print (k, l, m)

2 57 59


* 平均平方を求めてください。（上表 n, o）
  * 平均平方＝偏差平方和／自由度

In [17]:
# 平均平方
n = j / k
o = i / l
print (n, o)

374.3166666666672 30.62368421052631


* F値を求めてください。（上表 p ）
  * F　＝ 群間の平均平方／群内の平均平方

In [18]:
# F値
f_value = n / o
print (f_value)

12.223110194494591


これで分散分析表が完成しました。

<h4 style="border-bottom: solid 1px black;">5%有意水準のF分布表の値と比較する</h4>

これまで、カイ２乗検定を行うときは「カイ２乗分布表」を見ました。また、t検定のときは「t分布表」を見ました。

分散分析では「F分布表」で棄却域を見ます。

F分布表は次のようになっています。

<TABLE BORDER="1" align="center" CELLPADDING="2" CELLSPACING="0">
  <TR> 
    <TD WIDTH=100 rowspan="2" BGCOLOR="#CCCCCC"> <CENTER>
        群内の自由度 
      </CENTER>
      <div align="center"></div></TD>
    <TD colspan="5" BGCOLOR="#dddddd"> <CENTER>
        群間の自由度 
      </CENTER>
      <CENTER>
      </CENTER>
      <CENTER>
      </CENTER>
      <CENTER>
      </CENTER></TD>
  </TR>
  <TR> 
    <TD WIDTH=100 BGCOLOR="#dddddd"> <div align="center">1</div></TD>
    <TD WIDTH=100 BGCOLOR="#dddddd"> <div align="center">2</div></TD>
    <TD WIDTH=100 BGCOLOR="#dddddd"> <div align="center">3</div></TD>
    <TD WIDTH=100 BGCOLOR="#dddddd"> <div align="center">4</div></TD>
    <TD WIDTH=100 BGCOLOR="#dddddd"> <div align="center">5</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">10</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.96</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.10</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.71</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.48</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.33</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">20</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.35</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.49</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.10</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.87</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.71</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">30</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.17</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.32</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.92</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.69</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.53</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">40</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.08</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.23</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.84</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.61</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.45</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">50</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.03</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.18</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.79</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.56</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.40</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">60</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.00</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.15</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.76</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.53</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.37</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">70</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.98</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.13</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.74</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.50</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.35</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">80</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.96</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.11</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.72</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.49</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.33</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">90</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.95</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.10</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.71</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.47</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.32</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">100</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.94</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.09</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.70</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.46</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.31</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">200</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.89</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.04</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.65</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.42</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.26</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">300</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.87</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.03</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.63</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.40</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">2.24</div></TD>
  </TR>
</TABLE>

カイ２乗分布表も、t分布表も自由度はひとつでしたが、F分布表には自由度が2つあります。 群内の自由度と群間の自由度の両方を指定する必要があるからです。

今回は、群内の自由度が57ですので、一番近い60のところを見ます。さらに群間の自由度は2でしたので、2のところをみます。

そうすると、5%有意水準で、F=3.15が棄却域の境目であることがわかります。

いま、計算したFは12.22でしたので、5%有意水準で棄却域にはいります。

帰無仮説である「3つのお店のポテトの評価の平均に差はない」は棄却されました。

つまり「3つのお店のポテトの評価の平均には差がある」ということになります。

どれとどれの間に差があるかはわかりませんが、少なくともひとつの組み合わせの間で差があるということになります。

<h4 style="border-bottom: solid 1px black;">1%有意水準のF分布表の値と比較する</h4>

では、有意水準を1%としたときは、どうなるでしょうか。

下の表を使って、結論を下してみてください。

<TABLE BORDER="1" align="center" CELLPADDING="2" CELLSPACING="0">
  <TR> 
    <TD WIDTH=100 rowspan="2" BGCOLOR="#bCCCCC"> 
      <CENTER>
        群内の自由度 
      </CENTER>
      <div align="center"></div></TD>
    <TD colspan="5" BGCOLOR="#bddddd"> 
      <CENTER>
        群間の自由度 
      </CENTER>
      <CENTER>
      </CENTER>
      <CENTER>
      </CENTER>
      <CENTER>
      </CENTER></TD>
  </TR>
  <TR> 
    <TD WIDTH=100 BGCOLOR="#bddddd"> 
      <div align="center">1</div></TD>
    <TD WIDTH=100 BGCOLOR="#bddddd"> 
      <div align="center">2</div></TD>
    <TD WIDTH=100 BGCOLOR="#bddddd"> 
      <div align="center">3</div></TD>
    <TD WIDTH=100 BGCOLOR="#bddddd"> 
      <div align="center">4</div></TD>
    <TD WIDTH=100 BGCOLOR="#bddddd"> 
      <div align="center">5</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">10</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">10.04</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">7.56</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">6.55</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">5.99</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">5.64</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">20</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">8.10</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">5.85</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.94</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.43</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.10</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">30</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">7.56</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">5.39</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.51</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.02</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.70</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">40</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">7.31</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">5.18</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.31</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.83</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.51</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">50</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">7.17</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">5.06</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.20</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.72</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.41</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">60</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">7.08</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.98</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.13</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.65</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.34</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">70</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">7.01</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.92</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.07</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.60</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.29</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">80</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">6.96</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.88</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.04</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.56</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.26</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">90</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">6.93</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.85</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.01</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.53</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.23</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">100</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">6.90</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.82</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.98</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.51</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.21</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">200</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">6.76</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.71</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.88</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.41</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.11</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#bCCCCC"> 
      <div align="center">300</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">6.72</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">4.68</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.85</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.38</div></TD>
    <TD BGCOLOR="#eeeeee"><div align="right">3.08</div></TD>
  </TR>
</TABLE>

* 計算したF値と、F分布表を比較して、帰無仮説が棄却できるかどうか説明してください。

In [19]:
# 計算したF値と、F分布表を比較

# 群間の自由度は2、群内の自由度は57なので、5%有意水準のF値は約3.15、
# 同様に1%有意水準のF値は約4.98である。対して、計算したF値は12.2だった。
# したがって、どちらの有意水準においても棄却域に入るため、
# 帰無仮説である「3つのお店のポテトの評価に差はない」は棄却された。
# すなわち、ワクワク、モグモグ、パクパクのうち少なくともひとつの組み合わせの間で差があると言える。

<h4 style="border-bottom: solid 1px black;">多重比較</h4>

なお、どの組み合わせで差があるかということを、さらに調べるためには、多重比較という方法があります。

多重比較には、何種類かのやり方がありますが、この教材では扱いません。

<h2 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap6/sec4.html" target="_blank">6.4　実践編：3種類のハンバーガーの比較</a></h2>

「モグモグの味が変わった！」

<img src="image/m-burger00001.jpg">

店員さんが言いました。どうやらモグモグは、パクパクの進出を恐れ、ハンバーガーの改良を行ったらしいのです。

「パクパクも入れたハンバーガーの検定はやってないよね。モグモグも味を変えたことだし、3つで分散分析をやってよ」

<img src="image/salesman00009.jpg">

「おねがい！！」

手に15個ずつのハンバーガーを抱えあなたは駅に行き、データをとってきました。

それぞれ15人の女子高生、全部で45人の女子高生に食べてもらい、100点満点で点数をつけてもらいました。

<h4 style="padding: 0.25em 0.5em;color: #494949;background: transparent;border-left: solid 5px #7db4e6;">課題6.4</h4>

下記のデータを用いて、分散分析を行ってください。

In [20]:
wakuwaku = [90, 70, 75, 90, 65, 70, 80, 85, 75, 75, 85, 75, 80, 90, 80]
mogumogu = [75, 95, 75, 80, 75, 85, 75, 85, 80, 65, 90, 75, 90, 75, 85]
pakupaku = [65, 60, 75, 80, 65, 60, 70, 85, 70, 75, 75, 70, 80, 75, 80]

* まず帰無仮説と対立仮説を立てよう

In [21]:
# 帰無仮説 : ３つのグループの母集団の平均には差がない
# 対立仮説 : ３つのグループの母集団の少なくとも１つの組み合わせに差がある

* データ数、平均、分散、平方和を計算しよう

In [22]:
# データ数、平均、分散、平方和
print(len(wakuwaku), average(wakuwaku), variance(wakuwaku), sum_square_deviation(wakuwaku))
print(len(mogumogu), average(mogumogu), variance(mogumogu), sum_square_deviation(mogumogu))
print(len(pakupaku), average(pakupaku), variance(pakupaku), sum_square_deviation(pakupaku))

15 79.0 57.333333333333336 860.0
15 80.33333333333333 58.222222222222214 873.3333333333333
15 72.33333333333333 52.888888888888886 793.3333333333333


* 分散分析表を作ろう

In [23]:
# 分散分析表
zentai = wakuwaku + mogumogu + pakupaku
j = 0
for group in [wakuwaku, mogumogu, pakupaku]:
    j += (average(group) - average(zentai))**2 * len(group)

i = sum_square_deviation(wakuwaku) + sum_square_deviation(mogumogu) + sum_square_deviation(pakupaku)

h = sum_square_deviation(zentai)

k = 3 - 1
l = (len(wakuwaku) - 1) + (len(mogumogu) - 1) + (len(pakupaku) - 1)
m = len(wakuwaku + mogumogu + pakupaku) - 1

n = j / k
o = i / l

f_value = n / o

In [24]:
f_value

4.580474934036942

* F分布表を見て、分かることを書いてください。

In [25]:
# F分布表を見て、分かること

# 群間の自由度は2、群内の自由度は42なので、5%有意水準のF値は約3.23、
# 同様に1%有意水準のF値は約4.98である。対して、計算したF値は4.58だった。
# したがって、5%有意水準において棄却域に入るが、1%有意水準においては棄却域に入らない。
# 帰無仮説である「3つのお店のバーガーの評価に差はない」は5%有意水準において棄却された。
# すなわち、ワクワク、モグモグ、パクパクのうち少なくともひとつの組み合わせの間で、
# 5%有意水準において差があると言えるが、1%有意水準において差があると言えるほどの違いではない。

<h2 STYLE="background: #c2edff;padding: 0.5em;">ハンバーガー統計学 第６章 課題</h2>

__課題6.2__、および __課題6.4__ を解いて、指定のメールアドレスまでメールしてください。メール送信後は、エラーが帰ってきてないことを確認してください（メールアドレスを間違える人がときどき居ます）。
* 締切：10月17日（今すぐでなくても結構です）
* メールタイトル：「ハンバーガー統計学 第６章」
* 学籍番号と氏名を明記すること。
* 感想などがあれば書いてくれると嬉しいです。次回以降の講義の改善につながるかも知れません。

In [26]:
# ハンバーガー統計学 第６章 課題

<h2 STYLE="background: #c2edff;padding: 0.5em;">終わったら、<a href="HamburgerStatistics07.ipynb">第７章：分散分析（２要因）</a> に進んでください。</h2>
（2017年10月13日講義、10月17日課題提出締切）